File: Java/lesson 01/BinaryGap.java
// you can also use imports, for example:
// import java.util.*;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");

class Solution {
    public int solution(int N) {
        // write your code in Java SE 8
        int result = 0;
        int temp = -1;
        while (N != 0) {
            int x = N % 2;
            N = N / 2;
            if (temp >= 0 && x == 0) {
                temp++;
            } else if (x == 1) {
                if (temp < 0) {
                    temp = 0;
                } else {
                    result = result > temp? result: temp;
                    temp = 0;
                }
            }
        }
        return result;
    }
}

File: Go/lesson 01/BinaryGap.go
package solution

// you can also use imports, for example:
// import "fmt"
// import "os"

// you can write to stdout for debugging purposes, e.g.
// fmt.Println("this is a debug message")

func Solution(N int) int {
    // write your code in Go 1.4
    res := 0
    isStarted := false
    temp := 0
    for ; N != 0; N >>= 1 {
        if !isStarted {
            if N % 2 == 1 {
                isStarted = true
            }
        } else {
            if N % 2 == 1 {
                if temp > res {
                    res = temp
                }
                temp = 0
            } else {
                temp++
            }
        }
    }
    return res
}